ext/pdo_pgsql/pdo_pgsql.stub.php

```diff
@@ -52,4 +52,7 @@ public function lobUnlink(string $oid): bool {}
     public function getNotify(int $fetchMode = PDO::FETCH_DEFAULT, int $timeoutMilliseconds = 0): array|false {}
 
     public function getPid(): int {}
+
+    /** @implementation-alias PDO_PGSql_Ext::pgsqlSetNoticeCallback */
+    public function setNoticeCallback(?callable $callback): void {}
```

Comment:
Not going to lie, I think it might make more sense to _not_ magically add it to the PDO class and only have it on the new subclasses.

Reply:
My first reaction to your comment was a frank "yes, let's go!", but when starting to hack the code and adapt the tests,
I realized that as PDO::connect() is all new, _everybody_ still uses `new PDO(<url>)`; including the tests who use `PDOTest::factory()`, resulting in a `Call to undefined method [raw]PDO::setNoticeCallback()`.

Hardcoding the class seems to defeat the universal purpose of PDO (switching from SQLite to pg with a single DSN  change), so the only way would be to generalize `PDO::connect()` instead of `new PDO()`, starting with the tests.

But then, wouldn't it be the job of PDO to _first_ deprecate `new PDO()`, _then_ make everybody use it (particularly the frameworks that aim to give max access to the features with min hassle) so that it is sufficently mainstream that you know you'll benefit from all features of the driver without having to think about how to bootstrap your driver to gain access to them?

Reply:
Or I could switch the implementation to be the one of `PdopPgsql::setNoticeCallback()`, and make `PDO_PGSql_Ext::pgsqlSetNoticeCallback()` be the alias?
This would show our preference for future directions, while keeping the traditional way there for compatibility, while resolving the question of how to keep PDO instanciation both interface-generic and feature-specific and seeing how it gets adopted by third-party abstraction layers.
Just an idea, because I can think of no totally satisfying solution 😞

Reply:
The pdo subclass RFC recommends that driver-specific functionality be implemented in the subclass.

IMHO, I believe that implementing this in pdo core now would be contrary to that RFC.

So, I think if you want to implement this in pdo core, an RFC that overrides the subclasses will need to be accepted.

Reply:
> My first reaction to your comment was a frank "yes, let's go!", but when starting to hack the code and adapt the tests, I realized that as PDO::connect() is all new, _everybody_ still uses `new PDO(<url>)`; including the tests who use `PDOTest::factory()`, resulting in a `Call to undefined method [raw]PDO::setNoticeCallback()`.
> 
> Hardcoding the class seems to defeat the universal purpose of PDO (switching from SQLite to pg with a single DSN change), so the only way would be to generalize `PDO::connect()` instead of `new PDO()`, starting with the tests.
> 
> But then, wouldn't it be the job of PDO to _first_ deprecate `new PDO()`, _then_ make everybody use it (particularly the frameworks that aim to give max access to the features with min hassle) so that it is sufficently mainstream that you know you'll benefit from all features of the driver without having to think about how to bootstrap your driver to gain access to them?

The whole point of the subclassing RFC was to make Driver specific methods more robust, and phase out magically adding methods to the core PDO class.
Moreover, I don't see the issue with "forcing" users to migrate to the recommended way to be able to use a new feature.

Reply:
@Danack wrote:

> (it is possible you might get a more useful answer from someone not quite as in a mood as myself)

No worry! Having to regularly and tiringly argue at work against non-technical people that desperatly try to make it to the software architecture's hall of shame, with no will to do wrong but just an awful way of imagining how things work and thinking that everybody sticks to their desire,
I totally understand the effect my candid and insisting RFC-newbie questions can have.

> You're testing a DB specific method. It's fine to hard code the class name.
and @Girgias wrote:
> Moreover, I don't see the issue with "forcing" users to migrate to the recommended way to be able to use a new feature.

The concrete case I based my questions upon, is my paid job with 3 layers of software: PHP, SF/Doctrine, the app.
Each version upgrade is risky and costly (for tests), PHP and Doctrine upgrades sometimes get decorrelated.
If we upgraded to PHP 8.4 with a Doctrine that still uses the (_still supported_) `new PDO($dsn)` way to connect:
- with a magically added `PDO::pgsqlSetNoticeCallback()` the app could immediately benefit from it
- with only `PdoPgsql::setNoticeCallback()`, the app should wait for Doctrine to make the switch\
(admitedly this should come quick once `PDO::connect()` is published as the one best way to connect)

All is in the "_still supported_": as long as `new PDO` is valid, I would find regretful to add the feature to one entry point and not the other.
(I understand that it's a quick way to incite everybody to switch. But can everybody, as seen in my example?)

Reply:
I wrote:

> I would find regretful to add the feature to one entry point and not the other

And by the way, two other considerations:
- In fact I don't really like the DB-prefixed methods: I would have preferred a `PDO::setNoticeCallback()`.
As for `pgpsqlCopyFromArray()`: why not having named it `PDO::copyFromArray()` directly, thus creating a de-facto interface that other drivers could then choose to implement (or even rely on a default implementation in PDO, with parameterized `insert`s)
For this reason, **I find `PdoPgsql::setNoticeCallback()` vastly superior**.
- but I regret that this means limiting raw PDO usefulness ("Déshabiller Pierre pour habiller Paul", as we say in French: unclothe Peter to cloth Paul);
I like the hability to directly subclass PDO to add an applicative layer onto it _and_ benefit from the driver-added methods (as did Doctrine long time ago); I think of it as traits for DB drivers.
As I understand `PDO::connect()` hardcodes the instanciated class, so the only way to make a generic wrapper over PDO is now to make proxy containers that completely hide the PDO (as do Doctrine now), but make accessing driver-specific features a pain (unless you have enough developers to expose each driver-specific method in the corresponding proxy).

Reply:
On the other hand, I understand the "no full upgrade, no new feature!" motivation.

So with no "you get a point" for old, upgrade-illiterate people,
and given that even I would prefer a pure setNoticeCallback() (but on PDO),
yes, I may rework the whole thing to remove `pdopgsqlSetNoticeCallback()`.

(but thank you @nielsdos for the `@implementation-alias` that I keep for personal use 😄!)

Reply:
It was a fundamental PDO mistake to not have PDO be an abstract class/interface and require drivers to extend/implement it. If we are going to start adding magic methods to the base PDO class, then what was the point in the first place to introduce the dedicated driver classes?

The current design exists to preserve-BC by not randomly changing what class is returned by the constructor (which would be über weird). I think a long term plan is to deprecate instantiating a PDO object via `new` and force people to use the named constructor.

You are adding a new method, you are free to use the improved naming convention, it probably also makes sense to use the better naming on the dedicated driver classes. Especially as this hasn't been released yet.